### PR TITLE
Storage updates

### PR DESCRIPTION
**Pull request by dlongley**

### CODE REVIEW

File: lib/Collection.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+    if(sequence !== undefined && sequence !== doc.sequence) {
+      const err = new BedrockError(
+        'Could not update item; a conflict occurred.',
+        'InvalidStateError');
+      throw err;
+    }
+    const updatedDoc = {
+      ...doc
+    };
+    if(item) {
```
Comment (mattcollier) [inline lib/Collection.js]: Is it possible for `item` to be falsey here?  `item.type` is being accessed on L74.

File: lib/Collection.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+module.exports = class Collection {
+  constructor({type, instance, controllerKey, capability}) {
+    this.type = type;
+    this.instance = instance;
+    this.capability = capability;
+    this.controllerKey = controllerKey;
+  }
+
+  // FIXME: bikeshed name
```
Comment (mattcollier) [inline lib/Collection.js]: Is this still an open question?  What are appropriate alternatives?

File: lib/Collection.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+  static async getInstance({type, instance}) {
+    if(typeof instance === 'string') {
+      ({instance} = await instances.get({id: instance}));
+    }
+    const controllerKey = await instances.getControllerKey({id: instance.id});
+    const capability = await instances.getCapability(
+      {instance, referenceId: `${instance.id}-edv-configuration`});
```
Comment (mattcollier) [inline lib/Collection.js]: What's going on with the suffix `edv-configuration` here?

File: lib/Collection.js (hunk 2 above)
Comment (dlongley) [reply lib/Collection.js]: Yes, it's an open question and I don't know. `get`, `from`, `create` ... it's really a replacement for an async constructor. We can merge without addressing the FIXME, IMO.

File: lib/Collection.js (hunk 1 above)
Comment (dlongley) [reply lib/Collection.js]: Removed -- `item` must be given now.

File: lib/Collection.js (hunk 3 above)
Comment (dlongley) [reply lib/Collection.js]: It's the suffix used throughout the application that identifies the zcap associated with reading/write to the application's vault.

Comment (mattcollier) [reply lib/Collection.js]: Is there or would there be a collision on just the ID without the prefix?

Comment (dlongley) [reply lib/Collection.js]: @mattcollier, yes, there are many capabilities associated with an instance.

File: lib/Collection.js [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+    invocationSigner, client
+  });
+}
+
+async function getEdvClient({controllerKey, instance}) {
+  const [keyAgreementKey, hmac] = await Promise.all([
+    controllerKey.getKeyAgreementKey(instance.keys.kak),
+    controllerKey.getHmac(instance.keys.hmac)
+  ]);
+  // TODO: base this off of bedrock.config
```
Comment (mattcollier) [inline lib/Collection.js]: this is address by using: https://github.com/digitalbazaar/bedrock-https-agent

Comment (dlongley) [reply lib/Collection.js]: @mattcollier -- if you know how to fix, can you do so or file an issue for it?

File: lib/Collection.js [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  const invocationSigner = await controllerKey.getAsymmetricKey(
+    instance.keys.zcapKey);
+  const results = await client.find(
+    {equals, has, capability, invocationSigner});
+  return results;
+}
+
+// FIXME: make more restrictive, support `did:key` and `did:v1`
+async function keyResolver({id}) {
+  const httpsAgent = new https.Agent({
```
Comment (mattcollier) [inline lib/Collection.js]: https://github.com/digitalbazaar/bedrock-https-agent

File: lib/configurations.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
+/*!
+ * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+// load config defaults
+require('./config');
+
+const instances = require('./instances');
+// FIXME: use `Collection` instead
```
Comment (mattcollier) [inline lib/configurations.js]: How instead of what?

Comment (dlongley) [reply lib/configurations.js]: `walletClient.js` needs to be removed. It used to have the same stuff in it that `Collection` does now. We can't remove `walletClient.js` yet though -- as it has a couple of lines of code that need to be added to account creation that needs to get imported into this module from elsewhere. Once that's done, we can delete `walletClient.js`. This `configurations` stuff needs to be updated to use `Collection` in the meantime.

File: lib/http.js [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+
+  // get one or more issuer instances
+  app.get(
+    routes.instances,
+    ensureAuthenticated,
+    asyncHandler(async (req, res) => {
+      const {id, controller} = req.query;
+      if(id) {
+        const record = await vcIssuer.instances.get({id});
+        res.json(record);
```
Comment (mattcollier) [inline lib/http.js]: Return early.

File: lib/http.js [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+      const instance = {
+        id,
+        name,
+        controller
+      };
+      const record = await vcIssuer.instances.create({instance});
+      res.json(record);
+    }));
+
+  // get one or more issuer instances
```
Comment (mattcollier) [inline lib/http.js]: this is a bit atypical of our past with HTTP APIs, usually separate APIs for get one and getMany, is this a conscious decision.

File: lib/Collection.js (hunk 4 above)
Comment (mattcollier) [reply lib/Collection.js]: self-assigned issue

File: lib/instances.js [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+  await ensureKeystore({controllerKey});
+  return controllerKey;
+};
+
+// set issuer and capabilities (zcaps) for an instance
+exports.setIssuer = async ({id, controller, issuer, zcaps}) => {
+  const query = {'instance.id': id};
+  if(controller) {
+    // only allow update if controller matches
+    query.controller = {$in: [database.hash(controller)]};
```
Comment (mattcollier) [inline lib/instances.js]: Why is a new array being created here and `$in` being used?  Why not just direct compare?

File: lib/issuer.js [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+    {actor, accountId});
+  const issuerKey = await controllerKey.getAsymmetricKey({
+    id: invocationTarget.id,
+    type: invocationTarget.type,
+    capability
+  });
+
+  // create vcSigner API
+  const vcSigner = {
+    // TODO: need to be able to get the public key ID associated with the
```
Comment (mattcollier) [inline lib/issuer.js]: @dlongley can you turn this into an issue?  Option in github to create issue directly from this comment.

File: lib/zcaps.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
+/*!
+ * Copyright (c) 2019-2020 Digital Bazaar, Inc. All rights reserved.
+ */
+'use strict';
+
+// FIXME: use `base58-universal` instead
```
Comment (mattcollier) [inline lib/zcaps.js]: make issue.

File: lib/zcaps.js [hunk 12]
```diff
@@ -0,0 +1,215 @@
...
+      zcap.invocationTarget = {...parent.invocationTarget};
+      // delegate via externally referenced zcap key
+      delegated = await _delegate({zcap, signer});
+    }
+    results[delegated.referenceId] = delegated;
+  }
+  return results;
+};
+
+// FIXME: use did-key lib instead
```
Comment (mattcollier) [inline lib/zcaps.js]: does the library referenced exist?  If not, does it do much more than provide the document loader implemented here?

File: lib/configurations.js (hunk 6 above)
Comment (mattcollier) [reply lib/configurations.js]: @dlongley are you already doing this or do we need an issue?

File: lib/http.js (hunk 8 above)
Comment (dlongley) [reply lib/http.js]: Ah, I remember why this happened now. It used to be doing a look up by `issuer` but this changed recently during a big rewrite to just be an `id`. Previously, instances did not have IDs. So this should change now that they do and things work differently. Filed #8

File: lib/configurations.js (hunk 6 above)
Comment (dlongley) [reply lib/configurations.js]: Already have it marked on a punchlist.

File: lib/instances.js (hunk 9 above)
Comment (dlongley) [reply lib/instances.js]: It should be checking `controller` for an array/string value and doing a map call if it's an array (or normalizing to an array). Filed #9.

File: lib/issuer.js (hunk 10 above)
Comment (dlongley) [reply lib/issuer.js]: Filed #10.

**Pull request merged**